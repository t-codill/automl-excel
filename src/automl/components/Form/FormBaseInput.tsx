import * as React from "react";
import { Omit } from "../../common/Omit";
import { FormContext } from "./FormContext";
import { FormDataType } from "./FormDataType";
import { validate } from "./validate";
import { IValidator } from "./Validators";
export type FormBaseInputProps<
    TData extends { [key in TField]: TValue },
    TField extends keyof TData,
    TValue extends FormDataType,
    TProps> = Omit<TProps,
        "value"
        | "defaultValue"
        | "errorMessage"
        | "onChange"
        | "onChanged"
        | "defaultChecked"
        | "defaultValues"
        | "text"
        | "selectedKey"
        | "selectedKeys"
        | "defaultSelectedKey"
        | "defaultSelectedKeys"
        | "allowFreeform"
        | "validators"> & IFormBaseInputProps<TData, TField, TValue>;
export interface IFormBaseInputProps<
    TData extends { [key in TField]: TValue },
    TField extends keyof TData,
    TValue extends FormDataType
    > {
    field: TField;
    validators?: Array<IValidator<TValue>>;
    defaultFormValue?: TValue;
}
export interface IFormBaseInputStates<
    TData extends { [key in TField]: TValue },
    TField extends keyof TData,
    TValue extends FormDataType
    > {
    value: TValue;
    errorMessage: string | undefined;
}

export abstract class FormBaseInput<
    TData extends { [key in TField]: TValue },
    TField extends keyof TData,
    TValue extends FormDataType,
    TProps = {}
    >
    extends React.PureComponent<
    FormBaseInputProps<TData, TField, TValue, TProps>,
    IFormBaseInputStates<TData, TField, TValue>
    > {
    public static contextType = FormContext;

    public context!: React.ContextType<typeof FormContext>;

    constructor(props: FormBaseInputProps<TData, TField, TValue, TProps>) {
        super(props);
        this.state = {
            value: props.defaultFormValue as TValue,
            errorMessage: undefined
        };
    }

    public componentDidUpdate(prevProps: FormBaseInputProps<TData, TField, TValue, TProps>): void {
        if (prevProps.defaultFormValue !== this.props.defaultFormValue) {
            this.setState({ value: this.props.defaultFormValue as TValue });
        }
    }

    public componentDidMount(): void {
        this.context.mount(this);
    }

    public componentWillUnmount(): void {
        this.context.unmount(this);
    }

    public validate = () => {
        const error = validate<TValue>(this.state.value, this.props.validators);
        this.setState({ errorMessage: error });
        return error;
    }

    public getValue = (): TValue => {
        return this.state.value;
    }

    protected setValue(value: TValue): void {
        this.setState({
            value
        }, () => {
            if (this.context.onUpdated) {
                this.context.onUpdated(this.props.field, value);
            }
            this.validate();
        });
    }
}
