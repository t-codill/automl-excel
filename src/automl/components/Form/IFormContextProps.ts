import { IFormBaseInput } from "./IFormBaseInput";

export interface IFormContextProps {
    mount(component: IFormBaseInput): void;
    unmount(component: IFormBaseInput): void;
    // tslint:disable-next-line: no-any
    onUpdated?(key: any, value: any): void;
}
