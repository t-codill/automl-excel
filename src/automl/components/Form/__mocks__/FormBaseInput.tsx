import * as React from "react";
import { FormContext } from "../FormContext";

export abstract class FormBaseInput
    extends React.Component<{ defaultFormValue: object }, { value: object | undefined; errorMessage: undefined }> {
    public static contextType = FormContext;
    public context!: React.ContextType<typeof FormContext>;

    constructor(props: { defaultFormValue: object }) {
        super(props);
        this.state = {
            value: props.defaultFormValue,
            errorMessage: undefined
        };
    }

    public componentDidMount(): void {
        return;
    }

    public componentWillUnmount(): void {
        return;
    }

    public validate = () => {
        return;
    }

    public getValue = () => {
        return this.state.value;
    }

    protected setValue = (value: object) => {
        this.setState({
            value
        });
    }
}
