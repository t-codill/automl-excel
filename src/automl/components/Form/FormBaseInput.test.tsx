import { shallow, ShallowWrapper } from "enzyme";
import * as React from "react";
import { FormBaseInput } from "./FormBaseInput";
import { IFormContextProps } from "./IFormContextProps";

interface IFormBaseInputTestData {
  name: string;
}

const mount = jest.fn();
const unmount = jest.fn();
const onUpdated = jest.fn();
const validator = jest.fn();
const testFormContext: IFormContextProps = {
  mount,
  unmount,
  onUpdated
};

class FormBaseInputTester extends FormBaseInput<IFormBaseInputTestData, "name", string>{
  public get context(): IFormContextProps { return testFormContext; }
  public set context(_: IFormContextProps) { return; }
  public render(): React.ReactNode {
    return <div />;
  }
  public setValue(val: string): void {
    super.setValue(val);
  }
}

describe("FormBaseInput", () => {
  let wrapper: ShallowWrapper<{}, { value: string | undefined }, FormBaseInputTester>;
  beforeEach(() => {
    wrapper = shallow(
      <FormBaseInputTester
        field="name"
        defaultFormValue="defaultValue"
        validators={[validator]}
      />, {
        context: testFormContext
      });
  });

  it("should render", () => {
    expect(wrapper)
      .toMatchSnapshot();
  });

  it("should mount", (done) => {
    setImmediate(() => {
      expect(mount)
        .toBeCalledWith(wrapper.instance());
      done();
    });
  });

  it("should set default value", (done) => {
    wrapper.setProps({ defaultFormValue: "default value updated" });
    setImmediate(() => {
      expect(wrapper.state("value"))
        .toBe("default value updated");
      done();
    });
  });

  it("should validate", () => {
    const comp = wrapper
      .instance();
    comp.validate();
    expect(validator)
      .toBeCalledTimes(1);
    expect(comp.state.errorMessage)
      .toBeUndefined();
  });

  it("should validate with error", () => {
    const comp = wrapper
      .instance();
    validator.mockReturnValue("sample error");
    comp.validate();
    expect(validator)
      .toBeCalledTimes(1);
    expect(comp.state.errorMessage)
      .toBe("sample error");
  });

  it("should validate without validator", () => {
    const noValidateWrapper = shallow<FormBaseInputTester>(
      <FormBaseInputTester
        field="name"
        defaultFormValue="defaultValue"
      />, {
        context: testFormContext
      });
    const comp = noValidateWrapper
      .instance();
    comp.validate();
    expect(comp.state.errorMessage)
      .toBeUndefined();
  });

  it("should set and get value", () => {
    const comp = wrapper
      .instance();
    comp.setValue("test value");
    expect(comp.getValue())
      .toBe("test value");
    expect(onUpdated)
      .toBeCalledWith("name", "test value");
  });

  it("should set value without onUpdate callback", () => {
    const comp = wrapper
      .instance();
    testFormContext.onUpdated = undefined;
    comp.setValue("test value 2");
    expect(comp.getValue())
      .toBe("test value 2");
  });

  it("should unmount", () => {
    const comp = wrapper
      .instance();
    if (comp.componentWillUnmount) {
      comp.componentWillUnmount();
    }
    expect(unmount)
      .toBeCalledTimes(1);
  });
});
