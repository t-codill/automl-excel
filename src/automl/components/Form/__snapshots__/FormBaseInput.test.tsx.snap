// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormBaseInput should render 1`] = `<div />`;
